dbms/src/Storages/Page/V3/Universal/UniversalPageStorage.cpp

```diff
@@ -430,6 +430,15 @@ PS::V3::CPDataWriteStats UniversalPageStorage::dumpIncrementalCheckpoint(const U
     if (options.override_sequence)
         sequence = options.override_sequence.value();
 
+    {
+        // The output of `PageDirectory::dumpSnapshotToEdit` may contain page ids which are logically deleted but have not been gced yet.
+        // These page ids may be gcced when dump snapshot, so we cannot read data of these page ids.
+        // So we create a clean temp page_directory here and use it to dump edits with all visible page ids for `snap`.
+        PS::V3::universal::PageDirectoryFactory factory;
+        auto temp_page_directory = factory.dangerouslyCreateFromEditWithoutWAL(fmt::format("{}_{}", storage_name, sequence), edit_from_mem);
+        edit_from_mem = temp_page_directory->dumpSnapshotToEdit();
+    }
```

Comment:
This way is a bit wasteful of memory and CPU resources, and also increases the time consumed for uploading. Let's think about whether there is a better solution first.

Reply:
OK, I also hope for a better solution.

Reply:
There are no better solutions so far. Get this merge to fix the exception first.